external_services/api/views.py

```diff
@@ -0,0 +1,42 @@
+from rest_framework.permissions import IsAuthenticated
+from rest_framework.response import Response
+from rest_framework import status
+from rest_framework.views import APIView
+
+from .authentication import OAuthBodyHashAuthentication
+from .parsers import LTIOutcomeXMLParser
+from .renderers import LTIOutcomeResponseRenderer
+from .serializers import LTIOutcomeSerializer
+
+
+class LTIExerciseBasicOutcomesView(APIView):
+    '''View that responds to LTI 1.1 Basic Outcomes Service requests that
+    are posted by the Tool Provider to A+ in order to return the scores of submissions.
+    '''
+    
+    parser_classes = (LTIOutcomeXMLParser,)
+    authentication_classes = (OAuthBodyHashAuthentication,)
+    permission_classes = (IsAuthenticated,)
```

Comment:
Hmm.. So, any authenticated user that has authenticated using OAuthBodyHashAuthentication has access to write this resource? Thus, student using SessionAuthentication shouldn't be able to post here.

So, if anyone can get access to LTI secret, they can add submissions to any user to any exercise.

I think we should give some exercise specific secret when launching LTI service, so the service can only update that specific exercise with the secret.

Reply:
IP check (could be that is not be possible for LTI) - https://github.com/Aalto-LeTech/a-plus/blob/master/lib/api/authentication/grader.py#L28

Reply:
> I think we should give some exercise specific secret when launching LTI service, so the service can only update that specific exercise with the secret.

Now I remember, this was waiting for Aplus v2, where we will pre-create the submission, so we can have the secret for updating it. Currently, we don't have this secret to share with the grader.

Reply:
The LTI security is based on OAuth1 signature and body hash, which are based on the shared secret key. If the secret key is leaked, it should be changed. Knowing the secret allows one to post new submissions, indeed.

The LTI specification does not guarantee that the Tool Provider (TP) launch URL would correspond to the IP source address of the Basic Outcomes Service request when the TP posts grades back to A+. Therefore, I would not check the IP at all.

Reply:
Hmm. Yeah. So we go with that the LTI service is restricted to update submissions in exercises where it is linked (I forgot that restriction).

I Still would prefer that this endpoint would have permission that enforces the user is lti service.